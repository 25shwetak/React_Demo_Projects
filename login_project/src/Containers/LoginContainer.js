import React from "react";
import LoginComponent from "../Components/LoginComponent";
import * as yup from "yup";
import { useDispatch, useSelector } from "react-redux";
import { Redirect } from "react-router-dom";
import { setUsernameError, setPasswordError, resetErrors, setErrors, loginRequest} from "../actions/loginActions";

const LoginContainer = () => {
  //const [state, dispatch] = useReducer(loginReducer, initialState)

  //let {username, password} = state

  /*const [username, setUsername] = useState("")
  const [password, setPassword] = useState("")
  const [usernameError, setUsernameError] = useState(null)
  const [passwordError, setPasswordError] = useState(null)*/
  
  const dispatch = useDispatch();
  const result = useSelector((state) => state.loginDetailsReducer);
  const {
    username,
    password,
    usernameError,
    passwordError,
    userDetails,
  } = result;

  const schema = yup.object().shape({
        username: yup.string().email().required(),
        password: yup.string().min(6).required(),
      });

  const setUsernameErrorToNull = () => {
    dispatch(setUsernameError());
  }

  const setPasswordErrorToNull = () => {
    dispatch(setPasswordError());
  }

  // const validateData = () => {
  //   dispatch(resetErrors());
  //   schema.validate({ username, password }, { abortEarly: false })
  //   .then(() => {
  //     apiHelper('post', 'https://api.taiga.io/api/v1/auth', {username, password, type: "normal"})
  //     .then(({data}) => {
  //       dispatch(setUserDetails(data));
  //       console.log(data)
  //     })
  //   })
  //   .catch((err) => {
  //     err.inner.forEach((ele) => {
  //       dispatch(setErrors(ele));
  //     });
  //   });
  // }

  const validateData = () => {
    dispatch(resetErrors());
    schema.isValid({username, password})
    .then(function(valid) {
        if(!valid){
          schema.validate({ username, password }, { abortEarly: false })
          .catch((err) => {
            err.inner.forEach((ele) => {
                dispatch(setErrors(ele));
            });
          });
        } else {
            dispatch(loginRequest({username, password}));
        }
    })
  }

  if (userDetails.auth_token){
    return (<Redirect to="/dashboard" />)
  };

  return (
    <LoginComponent 
      validateData={validateData}
      usernameError={usernameError}
      passwordError={passwordError}
      username={username}
      password={password}
      dispatch={dispatch}
      setUsernameErrorToNull={setUsernameErrorToNull}
      setPasswordErrorToNull={setPasswordErrorToNull} />
  )
}

export default LoginContainer;