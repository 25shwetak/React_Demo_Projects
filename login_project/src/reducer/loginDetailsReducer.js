import { LOGIN_REDUCER } from "../shared/actionConstants";

export const initialState = {
  username: "",
  password: "",
  userDetails: {},
  usernameError: null,
  passwordError: null,
  request: {},
  loginFail: {}
};

const loginDetailsReducer = (state = initialState, action) => {
  switch (action.type) {
    case LOGIN_REDUCER.RESET_ERRORS:
      return { ...state, usernameError: null, passwordError: null }
    case LOGIN_REDUCER.SET_USERNAME:
      return { ...state, username: action.value };
    case LOGIN_REDUCER.SET_PASSWORD:
      return { ...state, password: action.value };
    case LOGIN_REDUCER.SET_USERNAME_ERROR:
      return { ...state, usernameError: action.value };
    case LOGIN_REDUCER.SET_PASSWORD_ERROR:
      return { ...state, passwordError: action.value };
    case LOGIN_REDUCER.SET_USER_DETAILS:
      return { ...state, userDetails: action.value };
    default:
      return state;
  }
};

export default loginDetailsReducer;