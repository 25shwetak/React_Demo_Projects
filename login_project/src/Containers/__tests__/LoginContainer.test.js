import{ render } from '@testing-library/react'
import { createStore, combineReducers } from 'redux';
import { Provider } from 'react-redux';
import React from "react";
import LoginContainer from "../LoginContainer";
import loginDetailsReducer from '../../reducer/loginDetailsReducer';
import projectsReducer from '../../reducer/projectsReducer';

const store = createStore(
  combineReducers({
    loginDetailsReducer, 
    projectsReducer
  })
);

describe("Login Container", () => {
  it("Must return login component", () => {
    const { asFragment } = render(
      <Provider store = {store}>
        <LoginContainer />
      </Provider>
    )

    expect(asFragment()).toMatchSnapshot();
  })
})