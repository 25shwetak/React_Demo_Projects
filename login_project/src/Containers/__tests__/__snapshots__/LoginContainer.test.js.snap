// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Container Must return login component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="m-auto col-sm-6 offset-sm-3"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <form
              class=""
            >
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="exampleEmail"
                >
                  Email
                </label>
                <input
                  aria-invalid="false"
                  class="form-control"
                  id="exampleEmail"
                  name="email"
                  placeholder="enter your email"
                  type="email"
                  value=""
                />
                <div
                  class="invalid-feedback"
                />
              </div>
              <div
                class="form-group"
              >
                <label
                  class=""
                  for="examplePassword"
                >
                  Password
                </label>
                <input
                  aria-invalid="false"
                  class="form-control"
                  id="examplePassword"
                  name="password"
                  placeholder="enter your password"
                  type="password"
                  value=""
                />
                <div
                  class="invalid-feedback"
                />
              </div>
            </form>
            <button
              class="btn btn-secondary"
              type="button"
            >
              Login
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
